lib/python/Pyxsim/testers.py

```diff
@@ -18,6 +19,123 @@ def __str__(self):
         return repr(self.value)
 
 
+class PytestComparisonTester:
```

Comment:
Is this a replacement for ComarisionTester?

Reply:
Sort of a replacement? I didn't want to replace an existing function with different functionality without knowing what else relies on this repo, and since some of the core features of ComparisonTester (writes to stderr on error, outputs results to stdout, returns the result) are different to features that make sense in a pytest realm (assert and break on error, reporting handled automatically) it ended up feeling like a very different function.

Reply:
I see, we assert a bit further up the chain for pytest. See https://github.com/xmos/lib_xud/blob/develop/tests/conftest.py

Im not sure which is the better approach without looking at it more carefully. However, I am vary of quite a lot of duplicated functionality. @kevinyeungxmos may have better insight. 

I would also like to keep this repo test-framework agnostic (not saying assessing necessary breaks this..)

Reply:
Hmm. We could pass a parameter (`use_assertions` or similar, defaulting to `False` to avoid breaking older code) that switches between the two behaviours (assert and break vs. using `record_failure` and continue)? A developer using `assert` should(?) identify that this will halt programme execution at point of fault and so shouldn't be surprised if this occurs - it then just so happens that Pytest overloads `assert` with its own hooks. The logical tests are equivalent and give the same information, it should just then be a case of choosing which set of logic we prefer and making sure the output is switchable.

Reply:
> Hmm. We could pass a parameter (use_assertions or similar, defaulting to False to avoid breaking older code) that switches between the two behaviours (assert and break vs. using record_failure and continue)? A developer using assert should(?) identify that this will halt programme execution at point of fault and so shouldn't be surprised if this occurs - it then just so happens that Pytest overloads assert with its own hooks. The logical tests are equivalent and give the same information, it should just then be a case of choosing which set of logic we prefer and making sure the output is switchable.

Agreed, however, I don't sure can the failure message show up properly using pytest with mult-proccessing. and also something like coverage test should do before comparison tester.

Reply:
> 

That makes sense - when you mention multiprocessing, do you mean forking before or after we call pytest? Pytest does support (with an extension) running tests on multiple cores - is this what you meant, or did you mean running multiple parallel instances of pytest?

I've not touched coverage yet - but yes, it would make sense that it would need to run before the comparison tester since, if we hit an assert, execution will stop. Is that the kind of note we could/should put in the documentation for this test framework?

Reply:
> That makes sense - when you mention multiprocessing, do you mean forking before or after we call pytest? Pytest does support (with an extension) running tests on multiple cores - is this what you meant, or did you mean running multiple parallel instances of pytest?

Yes, I mean pytest xdist. It allows running on multicore to speed up tests

> I've not touched coverage yet - but yes, it would make sense that it would need to run before the comparison tester since, if we hit an assert, execution will stop. Is that the kind of note we could/should put in the documentation for this test framework?

sure.

Reply:
The coverage tool needs to run *after* the comparison - we don't want to record coverage for failing tests.  This is one good case for asserting "higher up"